primitives/arithmetic/src/fixed.rs

```diff
@@ -0,0 +1,1308 @@
+				assert_eq!(a.into_inner(), inner_min);
+
+				let a = $name::saturating_from_rational(inner_min, -1);
+				assert_eq!(a.into_inner(), inner_max);
+
+				let a = $name::saturating_from_rational(inner_max, -1);
+				assert_eq!(a.into_inner(), inner_min);
+
+				let a = $name::saturating_from_rational(inner_max, inner_max);
+				assert_eq!(a.into_inner(), accuracy);
+
+				let a = $name::saturating_from_rational(inner_min, inner_min);
+				assert_eq!(a.into_inner(), accuracy);
+
+				let a = $name::saturating_from_rational(inner_max, -inner_max);
+				assert_eq!(a.into_inner(), -accuracy);
+
+				let a = $name::saturating_from_rational(-inner_max, inner_max);
+				assert_eq!(a.into_inner(), -accuracy);
+
+				let a = $name::saturating_from_rational(inner_max, 3 * accuracy);
+				assert_eq!(a.into_inner(), inner_max / 3);
+
+				let a = $name::saturating_from_rational(inner_max, -3 * accuracy);
+				assert_eq!(a.into_inner(), -inner_max / 3);
+
+				let a = $name::saturating_from_rational(inner_min, 2 * accuracy);
+				assert_eq!(a.into_inner(), inner_min / 2);
+
+				let a = $name::saturating_from_rational(inner_min, accuracy / -3);
+				assert_eq!(a.into_inner(), inner_max);
+
+				let a = $name::saturating_from_rational(inner_min, accuracy / 3);
+				assert_eq!(a.into_inner(), inner_min);
+
+				let a = $name::saturating_from_rational(1, accuracy);
+				assert_eq!(a.into_inner(), 1);
+
+				let a = $name::saturating_from_rational(1, -accuracy);
+				assert_eq!(a.into_inner(), -1);
+
+				let a = $name::saturating_from_rational(1, accuracy + 1);
+				assert_eq!(a.into_inner(), 0);
+
+				let a = $name::saturating_from_rational(1, -accuracy - 1);
+				assert_eq!(a.into_inner(), 0);
+			}
+
+			#[test]
+			fn checked_from_rational_works() {
+				let inner_max = <$name as FixedPointNumber>::Inner::max_value();
+				let inner_min = <$name as FixedPointNumber>::Inner::min_value();
+				let accuracy = $name::accuracy();
+
+				let a = $name::checked_from_rational(inner_max, accuracy).unwrap();
+				assert_eq!(a.into_inner(), inner_max);
+
+				let a = $name::checked_from_rational(inner_min, accuracy).unwrap();
+				assert_eq!(a.into_inner(), inner_min);
+
+				let a = $name::checked_from_rational(inner_max, -accuracy).unwrap();
+				assert_eq!(a.into_inner(), -inner_max);
+
+				let a = $name::checked_from_rational(inner_min, -accuracy);
+				assert_eq!(a, None);
+
+				let a = $name::checked_from_rational(inner_min + 1, -accuracy).unwrap();
+				assert_eq!(a.into_inner(), inner_max);
+
+				let a = $name::checked_from_rational(inner_max - 1, accuracy).unwrap();
+				assert_eq!(a.into_inner(), inner_max - 1);
+
+				let a = $name::checked_from_rational(inner_min + 1, accuracy).unwrap();
+				assert_eq!(a.into_inner(), inner_min + 1);
+
+				let a = $name::checked_from_rational(inner_max, 1);
+				assert_eq!(a, None);
+
+				let a = $name::checked_from_rational(inner_min, 1);
+				assert_eq!(a, None);
```

Comment:
I like where you are going here, but I want to make a correction. The first assertion in this test is basically the largerst value that this type can hold 

```
let a = $name::checked_from_rational(inner_max, accuracy).unwrap();
assert_eq!(a.into_inner(), inner_max);
```

and the next logical step to show that it is is to assert that `inner_max/(accuracy-1)` will be None, as it is an epsilon bigger than the previous one. 

and as per usual, comments are very important in my opinion and will save future you or anyone else a lot of time.